File: src/ens/lang/util/UnicodeScanner.java
package ens.lang.util;

import java.util.Arrays;

public class UnicodeScanner {
    public final static byte TAB = 0x9;
    public final static byte LF = 0xA;
    public final static byte FF = 0xC;
    public final static byte CR = 0xD;
    public final static byte EOI   = 0x1A;

    public char[] buffer;
    public int pointer;
    // Current char
    public char c;

    public int unicodeConversionPointer = -1;

    // Character buffer for saved chars.
    public char[] sbuffer = new char[128];
    public int realLength;
    public int spointer;

    public UnicodeScanner(char[] buffer) {
        this.buffer = Arrays.copyOf(buffer, buffer.length + 1);
        this.buffer[buffer.length] = EOI;
        pointer = -1;
        realLength = buffer.length;
        scanChar();
    }

    public char scanChar() {
        if (pointer < buffer.length) {
            c = buffer[++pointer];
            if (c == '\\') {
                convertUnicode();
            }
        }
        return c;
    }


    /**
     * Read next character in comment, skipping over double '\' characters.
     */
    public void scanCommentChar() {
        scanChar();
        if (c == '\\') {
            if (peekChar() == '\\' && !isUnicode()) {
                skipChar();
            } else {
                convertUnicode();
            }
        }
    }

    public boolean canScan() {
        return pointer < buffer.length;
    }

    public void putChar(char c, boolean scan) {
        if (spointer >= sbuffer.length) {
            sbuffer = Arrays.copyOf(sbuffer, spointer + 8);
        }
        sbuffer[spointer++] = c;
        if (scan)
            scanChar();
    }

    public void putChar(char c) {
        putChar(c, false);
    }

    public void putChar(boolean scan) {
        putChar(c, scan);
    }

    public void nextChar(boolean skip) {
        if (!skip) {
            if (spointer >= sbuffer.length) {
                sbuffer = Arrays.copyOf(sbuffer, spointer + 8);
            }
            sbuffer[spointer++] = c;
        }

        scanChar();
    }

    public String getSaved() {
        return new String(sbuffer, 0, spointer);
    }

    /**
     * Add 'count' copies of the character 'ch' to the string buffer.
     */
    public void repeat(char ch, int count) {
        for (; 0 < count; count--) {
            putChar(ch, false);
        }
    }

    /**
     * Reset the scan buffer pointer to 'pos'.
     */
    public void reset(int pos) {
        pointer = pos - 1;
        scanChar();
    }

    private void convertUnicode() {
        if (c == '\\' && unicodeConversionPointer != pointer) {
            pointer++;
            c = buffer[pointer];
            if (c == 'u') {
                do {
                    pointer++;
                    c = buffer[pointer];
                } while (c == 'u');
                int limit = pointer + 3;
                if (limit < buffer.length) {
                    int d = digit(pointer, 16);
                    int code = d;
                    while (pointer < limit && d >= 0) {
                        pointer++;
                        c = buffer[pointer];
                        d = digit(pointer, 16);
                        code = (code << 4) + d;
                    }
                    if (d >= 0) {
                        c = (char) code;
                        unicodeConversionPointer = pointer;
                        return;
                    }
                }
                System.err.println("Invalid Unicode escape, index: " + pointer);
            } else {
                pointer--;
                c = '\\';
            }
        }
    }

    public int peekSurrogates() {
        if (Character.isHighSurrogate(c)) {
            char high = c;
            int prevBP = pointer;

            scanChar();

            char low = c;

            c = high;
            pointer = prevBP;

            if (Character.isLowSurrogate(low)) {
                return Character.toCodePoint(high, low);
            }
        }

        return -1;
    }

    public int digit(int pos, int base) {
        char ch = c;
        if ('0' <= ch && ch <= '9')
            return Character.digit(ch, base); // a fast common case
        int codePoint = peekSurrogates();
        int result = codePoint >= 0 ? Character.digit(codePoint, base) : Character.digit(ch, base);
        if (result >= 0 && ch > 0x7f) {
            System.err.println("Illegal non-ascii digit, index: " + (pos + 1));
            if (codePoint >= 0)
                scanChar();
            c = "0123456789abcdef".charAt(result);
        }
        return result;
    }

    public boolean isUnicode() {
        return unicodeConversionPointer == pointer;
    }

    public void skipChar() {
        pointer++;
    }

    public char peekChar() {
        return buffer[pointer + 1];
    }
}

File: src/util/UnicodeScanner.cpp
#include "UnicodeScanner.h"

#include <iostream>

static int charDigit(int ch, int base) {
    int d;
    if ('0' <= ch && ch <= '9')      d = ch - '0';
    else if ('a' <= ch && ch <= 'z') d = ch - 'a' + 10;
    else if ('A' <= ch && ch <= 'Z') d = ch - 'A' + 10;
    else return -1;
    return d < base ? d : -1;
}

UnicodeScanner::UnicodeScanner(std::u16string_view src) {
    buffer.assign(src.begin(), src.end());
    buffer.push_back(EOI);
    realLength = (int)src.size();
    pointer = -1;
    unicodeConversionPointer = -1;
    sbuffer.resize(128);
    spointer = 0;
    c = 0;
    scanChar();
}

char16_t UnicodeScanner::scanChar() {
    if (pointer < (int)buffer.size()) {
        if (c == LF) {
            if (!prevWasCR) {
                line++;
            }
            column = 1;
            prevWasCR = false;
        } else if (c == CR) {
            line++;
            column = 1;
            prevWasCR = true;
        } else {
            column++;
            prevWasCR = false;
        }
        c = buffer[++pointer];
        if (c == u'\\') {
            convertUnicode();
        }
    }
    return c;
}

void UnicodeScanner::scanCommentChar() {
    scanChar();
    if (c == u'\\') {
        if (peekChar() == u'\\' && !isUnicode()) {
            skipChar();
        } else {
            convertUnicode();
        }
    }
}

bool UnicodeScanner::canScan() {
    return pointer < (int)buffer.size();
}

void UnicodeScanner::putChar(char16_t ch, bool scan) {
    if (spointer >= (int)sbuffer.size()) {
        sbuffer.resize(spointer + 8);
    }
    sbuffer[spointer++] = ch;
    if (scan)
        scanChar();
}

void UnicodeScanner::putChar(char16_t ch) {
    putChar(ch, false);
}

void UnicodeScanner::putChar(bool scan) {
    putChar(c, scan);
}

void UnicodeScanner::nextChar(bool skip) {
    if (!skip) {
        if (spointer >= (int)sbuffer.size()) {
            sbuffer.resize(spointer + 8);
        }
        sbuffer[spointer++] = c;
    }
    scanChar();
}

std::u16string UnicodeScanner::getSaved() const {
    return std::u16string(sbuffer.data(), spointer);
}

void UnicodeScanner::repeat(char16_t ch, int count) {
    for (; 0 < count; count--) {
        putChar(ch, false);
    }
}

void UnicodeScanner::reset(int pos) {
    pointer = pos - 1;
    scanChar();
}

int UnicodeScanner::digit(int pos, int base) {
    char16_t ch = c;
    if (u'0' <= ch && ch <= u'9')
        return charDigit(ch, base);
    int codePoint = peekSurrogates();
    int result = codePoint >= 0 ? charDigit(codePoint, base) : charDigit(ch, base);
    if (result >= 0 && ch > 0x7f) {
        std::cerr << "Illegal non-ascii digit, index: " << (pos + 1) << '\n';
        if (codePoint >= 0)
            scanChar();
        c = u"0123456789abcdef"[result];
    }
    return result;
}

bool UnicodeScanner::isUnicode() const {
    return unicodeConversionPointer == pointer;
}

void UnicodeScanner::skipChar() {
    pointer++;
}

char16_t UnicodeScanner::peekChar() const {
    return buffer[pointer + 1];
}

int UnicodeScanner::peekSurrogates() {
    if (c >= 0xD800 && c <= 0xDBFF) {
        char16_t high = c;
        int prevBP = pointer;
        scanChar();
        char16_t low = c;
        c = high;
        pointer = prevBP;
        if (low >= 0xDC00 && low <= 0xDFFF) {
            return (int)(((char32_t)(high - 0xD800) << 10) + (low - 0xDC00) + 0x10000);
        }
    }
    return -1;
}

void UnicodeScanner::convertUnicode() {
    if (c == u'\\' && unicodeConversionPointer != pointer) {
        pointer++;
        c = buffer[pointer];
        if (c == u'u') {
            do {
                pointer++;
                c = buffer[pointer];
            } while (c == u'u');
            int limit = pointer + 3;
            if (limit < (int)buffer.size()) {
                int d = charDigit(c, 16);
                int code = d;
                while (pointer < limit && d >= 0) {
                    pointer++;
                    c = buffer[pointer];
                    d = charDigit(c, 16);
                    code = (code << 4) + d;
                }
                if (d >= 0) {
                    c = (char16_t)code;
                    unicodeConversionPointer = pointer;
                    return;
                }
            }
            std::cerr << "Invalid Unicode escape, index: " << pointer << '\n';
        } else {
            pointer--;
            c = u'\\';
        }
    }
}
